vllm/model_executor/guided_decoding.py

```diff
@@ -0,0 +1,166 @@
+import asyncio
+from collections import defaultdict
+import concurrent.futures
+from copy import copy
+from enum import Enum
+from functools import lru_cache
+from json import dumps as json_dumps
+from re import escape as regex_escape
+from typing import Union
+from types import SimpleNamespace
+from pydantic import BaseModel
+
+from vllm.entrypoints.openai.protocol import CompletionRequest, ChatCompletionRequest
+try:
+    from outlines.serve.vllm import JSONLogitsProcessor, RegexLogitsProcessor
+except ImportError as e:
+    raise ValueError(
+        "Please install 'outlines' (pip install outlines) to use guided decoding."
+    ) from e
+
+
+class GuidedDecodingMode(Enum):
+    JSON = "json"
+    REGEX = "regex"
+    CHOICE = "choice"
+    GRAMMAR = "grammar"
```

Comment:
**suggestion**

We have added the grammar mode, but there is still no logits processor for grammar guided decoding in outlines (cf. https://github.com/outlines-dev/outlines/blob/main/outlines/serve/vllm.py). While it shouldn't be too difficult to add it there, this will take at least a few days. Maybe we should remove this from the enum and add it in a later PR to avoid bloating this PR? @simon-mo, what do you think?

Reply:
I've added grammar support for now by working off of this [outlines PR](https://github.com/outlines-dev/outlines/pull/541), simon said it should be merged very soon

Reply:
okay!

Reply:
Hey @rlouf, I've copied the logits processor code from outlines over into this PR. Let me know if there's anything you'd like to change to the new file (guided_logits_processors.py).

Reply:
LGTM